src/daemon/daemon_rpc.cpp

```diff
@@ -92,143 +100,219 @@ grpc::Status emit_signal_and_wait_for_result(OperationSignal operation_signal)
-        std::bind(&DaemonRpc::on_delete, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_delete, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::umount(grpc::ServerContext* context, const UmountRequest* request,
                                    grpc::ServerWriter<UmountReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_umount, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_umount, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::version(grpc::ServerContext* context, const VersionRequest* request,
                                     grpc::ServerWriter<VersionReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_version, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_version, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::ping(grpc::ServerContext* context, const PingRequest* request, PingReply* response)
 {
+    if (connection_type == mp::RpcConnectionType::ssl)
+    {
+        auto client_cert = client_cert_from(context);
+
+        if (!client_cert.empty() && client_cert_store->verify_cert(client_cert))
+        {
+            return grpc::Status::OK;
+        }
+
+        return grpc::Status{grpc::StatusCode::UNAUTHENTICATED, ""};
+    }
+
     return grpc::Status::OK;
 }
 
 grpc::Status mp::DaemonRpc::get(grpc::ServerContext* context, const GetRequest* request,
                                 grpc::ServerWriter<GetReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_get, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_get, this, request, response, std::placeholders::_1), client_cert_from(context));
+}
+
+grpc::Status mp::DaemonRpc::authenticate(grpc::ServerContext* context, const AuthenticateRequest* request,
+                                         grpc::ServerWriter<AuthenticateReply>* response)
+{
+    auto status = emit_signal_and_wait_for_result(
+        std::bind(&DaemonRpc::on_authenticate, this, request, response, std::placeholders::_1));
+
+    if (status.ok() && connection_type == mp::RpcConnectionType::ssl)
+    {
+        const auto store_was_empty{client_cert_store->is_store_empty()};
+
+        client_cert_store->add_cert(client_cert_from(context));
+
+        if (store_was_empty && server_socket_type == mp::ServerSocketType::unix)
+        {
+            MP_PLATFORM.set_server_permissions(server_address, false);
+        }
```

Comment:
Kinda feel like we should defer to the platform when / how to restrict? So you wouldn't need `store_was_empty` or `server_socket_type` at all. I mean, it should be idempotent, so we can call it on start and then every time someone authenticates.

Reply:
Well, this is tricky.  We only care about Unix sockets and setting the proper permissions on this socket depending on if the store is empty to allow the first client to connect on the Unix socket to have its cert automatically imported.  So really, this is a function of the socket type and the contents of the client cert store.  I'll think about how I can make this better, but this took some iterations already to get to where this is now.

Reply:
Ok, I cleaned this up. Lemme know what you think.

Reply:
This has changed quite a bit since these comments.  I'm resolving this conversation and a new conversation can be started with the newer code in place.